_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+
+We're working on another blog post that digs deeper into how we use `styled-system` within
+[our design system](https://palette.artsy.net/). You can also
+[poke around our source](https://github.com/artsy/reaction/blob/f3dabb884d616c5c42bbd47b1432dc2a9456b8ca/src/Apps/Search/Components/SearchResultsSkeleton/Header.tsx#L6)
+to see how much we've embraced `styled-system`'s responsive styles. (In that example, `pr` means `padding-right`
+and `pl` means `padding-left`.)
+
+There's one type of challenge with building a responsive app that `styled-system` can't solve: when we need to emit
+different layouts across different breakpoints. In this case, we need something that can render very different
+component sub-trees. We couldn't find an approach that satisfied our needs, so we wrote our own.
+
+## Tool 2: [`@artsy/react-responsive-media`](https://github.com/artsy/react-responsive-media)
+
+First off, an announcement: we've just released `react-responsive-media` version 1.0! (TODO: release v1.0 😬 or
+change this announcement)
+
+`react-responsive-media` allows you to define a set of breakpoint widths, then declaratively render component
+sub-trees when those breakpoints are met. It looks something like this:
+
+```xml
+<>
+  <Media at="xs">
+    <MobileLayout />
+  </Media>
+  <Media greaterThan="xs">
+    <NonMobileLayout />
+  </Media>
+</>
+```
+
+In this example, we're emitting the `MobileLayout` component for devices at or below our `xs` breakpoint, and the
+`NonMobileLayout` for devices greater than our `xs` breakpoint. You can imagine that the `MobileLayout` and
+`NonMobileLayout` components are complicated sub-trees, with more significant differences than `styled-system`
+could handle.
+
+### How it works
+
+The first important thing to note is that when server-rendering with `react-responsive-media`, **all** breakpoints
+get rendered by the server.
+
+Why not just the one that the current device needs? A couple reasons. First, we can't _accurately_ identify which
+breakpoint your device needs from the server. We could use a library to sniff the browser `user-agent`, but those
+aren't always accurate, and they wouldn't give us all the information we need to know when we are server-rendering.
+
+The other reason `react-responsive-media` renders _all_ breakpoints has to do with the hydration of a
+server-rendered React app. When using SSR in a React app, you need to call
+[`ReactDOM.hydrate()`](https://reactjs.org/docs/react-dom.html#hydrate) to connect your server-rendered app to the
+client's browser. As the docs describe,
+
+> React expects that the rendered content is identical between the server and the client.
+
+If there are differences between your server-rendered component tree and the initial client-rendered component
+tree, a couple things will happen.
+
+1. You'll get a friendly error in your console telling you there were differences (in development mode), but
+   also...
+2. weird, unpredictable things will happen to your app.
```

Comment:
Not "will", but "can". React's hydration mechanism isn't fragile, and there's even a setting -- `suppressHydrationWarning` -- to silence it, because there are cases where mismatches can't be avoided or are simply harmless.

Reply:
> React's hydration mechanism isn't fragile

Mmmmm... I don't know that I agree with that. It can break transparently just based off of the wrapping elements you use in a component. `suppressHydrationWarning` does give a mechanism to suppress the warning, but it doesn't reduce the risk. Sure it _may_ be harmless in a lot of cases, but it also spawn hard to debug failures when people make innocuous changes like adding a wrapping div to a component that just so happens to match the same order of rendered output from the server. 

I generally think React 15's approach of tossing out the missmatched DOM tree was _much_ safer if less performant, but I guess that's neither here nor there.

Reply:
I guess my point is more along the lines of -- is this a _separate_ blog post, or does it belong here, in our post about RRM? They seem to be only tangentially related, and unrelated to the bizarre hydration issue we've seen.

Reply:
I reworded some things here. https://github.com/artsy/artsy.github.io/pull/566/files#diff-a879e573d311218fd7eec7da14d776f5R107 and https://github.com/artsy/artsy.github.io/pull/566/files#diff-a879e573d311218fd7eec7da14d776f5R113.